File: npm/cli-with-claude-skill/skill/template/python/skill_install.py
"""Drop into your click-based CLI.

Project layout:
    mytool/skill.md            (the embedded skill markdown — packaged as data)
    mytool/skill_install.py    (this file)
    mytool/__main__.py         (your click entrypoint)

Wire-up in __main__.py:

    import click
    from .skill_install import install_claude_skill_cmd, check_skill_notice

    @click.group()
    def cli():
        check_skill_notice()  # startup hook

    cli.add_command(install_claude_skill_cmd)
    # ...your other subcommands

    if __name__ == "__main__":
        cli()

Packaging note: ensure ``skill.md`` is included as package data so
``importlib.resources`` can find it. In ``pyproject.toml`` (hatchling/setuptools):

    [tool.hatch.build.targets.wheel]
    packages = ["mytool"]
    [tool.hatch.build.targets.wheel.force-include]
    "mytool/skill.md" = "mytool/skill.md"

Replace ``mytool`` everywhere with your actual binary / package name.
"""

from __future__ import annotations

import os
from importlib import resources
from pathlib import Path

import click

TOOL_NAME = "mytool"


def _load_embedded() -> str:
    return resources.files(__package__).joinpath("skill.md").read_text(encoding="utf-8")


EMBEDDED: str = _load_embedded()


def skill_path() -> Path:
    return Path.home() / ".claude" / "skills" / TOOL_NAME / "SKILL.md"


def parse_version(content: str) -> str | None:
    for block in content.split("---", 2):
        for line in block.splitlines():
            stripped = line.strip()
            if stripped.startswith("version:"):
                return stripped[len("version:") :].strip()
    return None


def parse_semver(v: str) -> tuple[int, int, int] | None:
    s = v.strip()
    if s.startswith("v"):
        s = s[1:]
    for sep in ("-", "+"):
        if sep in s:
            s = s.split(sep, 1)[0]
    parts = s.split(".")
    if not parts or len(parts) > 3:
        return None
    out = [0, 0, 0]
    for i, p in enumerate(parts):
        if not p.isdigit():
            return None
        out[i] = int(p)
    return out[0], out[1], out[2]


def should_show_outdated(installed: str | None, embedded: str | None) -> bool:
    if not embedded:
        return False
    if not installed:
        return True
    a = parse_semver(installed)
    b = parse_semver(embedded)
    if a is None or b is None:
        return installed != embedded
    return b > a


def install_claude_skill() -> None:
    path = skill_path()
    path.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    path.write_text(EMBEDDED, encoding="utf-8")
    click.echo(f"Claude Code skill installed at {path}")


def check_skill_notice() -> None:
    path = skill_path()
    try:
        installed = path.read_text(encoding="utf-8")
    except FileNotFoundError:
        click.echo(
            f"Tip: Run `{TOOL_NAME} install-claude-skill` to add Claude Code skill support.",
            err=True,
        )
        return
    if should_show_outdated(parse_version(installed), parse_version(EMBEDDED)):
        click.echo("Notice: Claude skill is outdated. Auto-updating...", err=True)
        try:
            install_claude_skill()
        except OSError as e:
            click.echo(
                f"  failed: {e}. Run `{TOOL_NAME} install-claude-skill` to update manually.",
                err=True,
            )


@click.command(name="install-claude-skill")
def install_claude_skill_cmd() -> None:
    """Install/update the Claude Code skill into ~/.claude/skills/<tool>/."""
    install_claude_skill()

File: npm/cli-with-claude-skill/skill/template/rust/skill_install.rs
//! Drop into your clap-based CLI.
//!
//! Project layout:
//!   src/skill.md            (the embedded skill markdown)
//!   src/skill_install.rs    (this file)
//!   src/main.rs             (your clap derive entrypoint)
//!
//! Wire-up in main.rs:
//!
//!   mod skill_install;
//!   use clap::{Parser, Subcommand};
//!
//!   #[derive(Parser)]
//!   #[command(name = "mytool")]
//!   struct Cli { #[command(subcommand)] command: Commands }
//!
//!   #[derive(Subcommand)]
//!   enum Commands {
//!       /// Install/update the Claude Code skill into ~/.claude/skills/mytool/
//!       InstallClaudeSkill,
//!       // ...your other subcommands
//!   }
//!
//!   fn main() {
//!       skill_install::check_notice();          // startup hook (missing → tip; outdated → auto-update)
//!       let cli = Cli::parse();
//!       match cli.command {
//!           Commands::InstallClaudeSkill => skill_install::install().unwrap(),
//!           // ...
//!       }
//!   }
//!
//! Replace `mytool` everywhere with your actual binary name.
//!
//! Cargo.toml dependency: `dirs = "5"` (for home-dir lookup).

use std::fs;
use std::io::Write;
use std::path::PathBuf;

pub const TOOL_NAME: &str = "mytool";

pub const EMBEDDED: &str = include_str!("skill.md");

pub fn skill_path() -> PathBuf {
    let home = dirs::home_dir().expect("home dir");
    home.join(".claude").join("skills").join(TOOL_NAME).join("SKILL.md")
}

/// Returns the value after `version:` in the YAML frontmatter, or None.
pub fn parse_version(content: &str) -> Option<String> {
    for block in content.splitn(3, "---") {
        for line in block.lines() {
            let line = line.trim();
            if let Some(rest) = line.strip_prefix("version:") {
                return Some(rest.trim().to_string());
            }
        }
    }
    None
}

/// Parse `[major, minor, patch]` from a version string. Accepts `1`, `1.2`, `1.2.3`, optional `v` prefix.
pub fn parse_semver(v: &str) -> Option<[u32; 3]> {
    let v = v.trim();
    let v = v.strip_prefix('v').unwrap_or(v);
    let v = v.split(|c| c == '-' || c == '+').next()?;
    let parts: Vec<&str> = v.split('.').collect();
    if parts.is_empty() || parts.len() > 3 {
        return None;
    }
    let mut out = [0u32; 3];
    for (i, p) in parts.iter().enumerate() {
        out[i] = p.parse().ok()?;
    }
    Some(out)
}

pub fn should_show_outdated(installed: Option<&str>, embedded: Option<&str>) -> bool {
    let Some(emb) = embedded else { return false };
    let Some(inst) = installed else { return true };
    match (parse_semver(inst), parse_semver(emb)) {
        (Some(a), Some(b)) => b > a,
        _ => inst != emb,
    }
}

pub fn check_notice() {
    let path = skill_path();
    let installed = match fs::read_to_string(&path) {
        Ok(s) => s,
        Err(_) => {
            eprintln!("Tip: Run `{} install-claude-skill` to add Claude Code skill support.", TOOL_NAME);
            return;
        }
    };
    let installed_v = parse_version(&installed);
    let embedded_v = parse_version(EMBEDDED);
    if should_show_outdated(installed_v.as_deref(), embedded_v.as_deref()) {
        eprintln!("Notice: Claude skill is outdated. Auto-updating...");
        if let Err(e) = install() {
            eprintln!("  failed: {}. Run `{} install-claude-skill` to update manually.", e, TOOL_NAME);
        }
    }
}

pub fn install() -> std::io::Result<()> {
    let path = skill_path();
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let mut f = fs::File::create(&path)?;
    f.write_all(EMBEDDED.as_bytes())?;
    println!("Claude Code skill installed at {}", path.display());
    Ok(())
}
